index.js

```diff
@@ -88,11 +88,23 @@ function fastifyWebsocket (fastify, opts, next) {
   // to monkeypatching for now.
   const oldClose = fastify.server.close
   fastify.server.close = function (cb) {
-    const server = fastify.websocketServer
-    for (const client of server.clients) {
-      client.close()
-    }
+    // Call oldClose first so that we stop listening. This ensures the
+    // server.clients list will be up to date when we start closing below.
     oldClose.call(this, cb)
+
+    function closeAllClients () {
+      if (fastify.server.listening) {
+        setTimeout(closeAllClients, 0)
+        return
+      }
```

Comment:
This is the best approach I could come up with, @mcollina, although I'm not super happy with it.

It was a bit of a dance to coordinate the exact sequence of events with test, but I managed to do it by by adding this `setTimeout`, which allows the websocket client created in the test to connect. After thinking about it, this isn't a bad idea outside of test. We want to avoid the `for` loop below until the server is no longer listening.

Let me know what you think.

Reply:
Note: I verified that the original code would result in a test timeout, so I'm quite confident the test is capturing the essence of what we're trying to achieve with this PR.

Reply:
This looks fine to me.

A possible alternative could be to maybe update the `handleRouting` function to close incoming connections if the fastify instance is in `closing` state? 
I haven't given it deeper thought though, so I'm not sure if I missed something

Reply:
@mcollina I was also wary at first, but we need to provide the guarantee that `fastify.server.listening === false` at this point, otherwise we'll still have our problematic race condition. Stubbing out `fastify.server` would likely just result in us creating the a sequence of events that doesn't actually capture the real server.

The thing I do like about the `setTimeout` approach is that it provides that guarantee without assuming anything about the underlying implementation of `fastify` or `fastify.server`, which I think is an important detail of this PR (since one can provide a custom server impl).

> A possible alternative could be to maybe update the `handleRouting` function to close incoming connections if the fastify instance is in `closing` state?

I completely missed that function, @AyoubElk. Thanks for pointing it out! I'll play around with it to see if I can achieve the same result. Looks like all of that state is likely set before we'd get to this point (https://github.com/fastify/fastify/blob/be4ec2d193e38ffd93a786f808e516374eb46597/fastify.js#L357-L358), so there's a good chance it'll work!

Reply:
Ok! Could you use a setImmediate instead?

Reply:
I'll try @AyoubElk's idea first, because I do like the idea of the just closing incoming websocket connections (and we can close them with a 1001 too, to ensure a cleaner, more informative close).

If I can't get that working at all, I'll fall back to `setImmediate`. If any other ideas come to mind in the meantime, do let me know. I tried a lot of different things before settling on this solution.